### PR TITLE
Add persistence callbacks

### PR DESCRIPTION
**Pull request by richmolj**
These allow us to avoid overriding create/update directly and instead
rely on hooks:

* `before_attributes`: before the request attributes are assigned to the
model. Accepts attributes as argument.
* `after_attributes`: after the attributes have been assigned to the
but before save. Accepts the model as argument.
* `around_attributes`: Same as the above, but you must register with a
method name and `yield attributes` within it. Accepts attributes as
argument and the model instance is result of yield.
* `before_save`: after attributes have been assigned but before save.
Accepts model as argument.
* `after_save`: after the object has been persisted, but before that
transaction has been committed.
* `around_save`: Same as above, accepts the unpersisted model instance
and the persisted model instance is the result of yield.
* `before/after/around_destroy`: Same as all the above, accepts the
model instance.

See specs for details.

### CODE REVIEW

**Comment by richmolj** [comment]
@wadetandy please take a look at this one

File: lib/graphiti/resource/configuration.rb [hunk 1]
```diff
@@ -169,7 +169,8 @@ def config
               before_commit: {},
               attributes: {},
               extra_attributes: {},
-              sideloads: {}
+              sideloads: {},
```
Comment (wadetandy) [inline lib/graphiti/resource/configuration.rb:172]: This is starting to get a bit big to be a hash, IMO.  Doesn't need to be here but I think the next time we need to add to this it should be turned into an actual object (or series of objects, depending on nested complexity).

**Review by wadetandy — APPROVED** [review]
Small comment on future maintainability but feature looks great.

File: lib/graphiti/resource/configuration.rb (hunk 1 above)
Comment (richmolj) [reply lib/graphiti/resource/configuration.rb:172]: Don't disagree but the biggest issue is getting inheritance and overrides to play correctly. Just a note.

**Comment by wadetandy** [comment]
When using our generators for non-activerecord models, we already create a test which has a factory that likely won't exist, and it's up to the user to replace it with the correct logic.  The generators are a tool to get up and running quickly and can't possibly foresee every use case that someone might have. Perhaps we could later come up with a smooth way to handle your edge case, but I'd generally say this is beyond the scope of helping with a simple attributes payload.

**Pull request merged**